db/cql_type_parser.cc

```diff
@@ -93,58 +95,31 @@ class db::cql_type_parser::raw_builder::impl {
         /*
          * build a DAG of UDT dependencies
          */
-        std::unordered_multimap<entry *, entry *> adjacency;
+        std::vector<entry *> all_definitions;
+        std::multimap<entry *, entry *> adjacency;
         for (auto& e1 : _definitions) {
+            all_definitions.emplace_back(&e1);
             for (auto& e2 : _definitions) {
                 if (&e1 != &e2 && std::any_of(e1.field_types.begin(), e1.field_types.end(), [&e2](auto& t) { return t->references_user_type(e2.name); })) {
                     adjacency.emplace(&e2, &e1);
                 }
             }
-        }
-        /*
-         * resolve dependencies in topological order, using Kahn's algorithm
-         */
-        std::unordered_map<entry *, int32_t> vertices; // map values are numbers of referenced types
-        for (auto&p : adjacency) {
-            vertices[p.second]++;
-        }
-
-        std::deque<entry *> resolvable_types;
-        for (auto& e : _definitions) {
-            if (!vertices.contains(&e)) {
-                resolvable_types.emplace_back(&e);
-            }
+            co_await coroutine::maybe_yield();
         }
 
         // Create a copy of the existing types, so that we don't
         // modify the one in the keyspace. It is up to the caller to
         // do that.
         replica::user_types_metadata types = _ks.user_types();
-
         const auto &ks_name = _ks.name();
-        std::vector<user_type> created;
 
-        while (!resolvable_types.empty()) {
-            auto* e =  resolvable_types.front();
-            auto r = adjacency.equal_range(e);
-
-            while (r.first != r.second) {
-                auto* d = r.first->second;
-                if (--vertices[d] == 0) {
-                    resolvable_types.push_back(d);
-                }
-                ++r.first;
-            }
+        auto sorted = co_await utils::topological_sort(all_definitions, adjacency);
+        std::vector<user_type> created;
+        created.reserve(sorted.size());
 
+        for (auto* e : sorted) {
             created.push_back(e->prepare(ks_name, types));
-            types.add_type(created.back());
-            resolvable_types.pop_front();
         }
```

Comment:
No, the new logic is wrong. You are missing the `types.add_type(created.back());` line. Without it, the new definitions of dependent types are going to reference the *old* versions of the dependees.

Which misses the entire point of this function, which is to recreate the dependents with updated references.

If tests don't catch this, then this code has no proper tests.

@piodul @Jadw1 I apologize for suggesting to fix what isn't broken. Fear the old code.

Reply:
Let's pretend we never noticed the old code, lest we break something.

Reply:
No, let's add a test that passes with the old code, fails with this series, and passes again after a fix

Reply:
I think that the test case `test_udts_with_dependencies` that we added was supposed to take care of that, but it should be improved.